src/artifacts.js

```diff
@@ -423,6 +643,119 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
   this.monitor.reportError(err);
 };
 
+/** Complete artifact */
+api.declare({
+  method:     'put',
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'completeArtifact',
+  stability:  API.stability.experimental,
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>',
+    ], [
+      'queue:create-artifact:<taskId>/<runId>',
+    ],
+  ],
+  deferAuth:  true,
+  input:      'put-artifact-request.json#',
+  output:     'put-artifact-response.json#',
+  title:      'Complete Artifact',
+  description: 'tbd',
+}, async function(req, res) {
+  let taskId      = req.params.taskId;
+  let runId       = parseInt(req.params.runId, 10);
+  let name        = req.params.name;
+  let input       = req.body;
+  let isPublic    = /^public\//.test(name);
+
+  // Hmm, list destructuring is probably better here
+  let [artifact, task] = await Promise.all([
+    this.Artifact.load({taskId, runId, name}, true),
+    this.Task.load({taskId}, true),
+  ]);
+
+  // TODO: Have a shortcut here where things which are already
+  // marked as present short circuit
+
+  let run = task.runs[runId];
+  if (!run) {
+    return res.reportError('InputError',
+      'Run not found',
+      {});
+  }
+  let workerGroup = run.workerGroup;
+  let workerId = run.workerId;
+
+  // Authenticate request by providing parameters
+  if (!req.satisfies({
+    taskId,
+    runId,
+    workerGroup,
+    workerId,
+    name,
+  })) {
+    return;
+  }
+
+  if (artifact.storageType === 'blob') {
+    let etag;
+    if (artifact.details.partsHash) {
+      etag = await this.s3Controller.completeMultipartUpload({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        etags: input.etags,
+        uploadId: artifact.details.uploadId,
+        tags: {taskId, runId, name},
+      });
+    } else {
+      let url = await this.s3Controller.generateGetUrl({
+        bucket: artifact.details.bucket,
+        key: artifact.details.key,
+        signed: !isPublic,
+      });
+
+      if (urllib.parse(url).protocol !== 'https:') {
+        throw new Error('Only HTTPS is allowed for artifacts');
+      }
+
+      let headRes = await this.s3Runner.run({
+        req:{
+          url,
+          method: 'HEAD',
+          headers: {},
+        },
+      });
+      
+      if (headRes.headers['x-amz-meta-content-sha256'] !== artifact.details.contentSha256) {
+        throw new Error('S3 object does not have correct Content-Sha256 metadata');
+      }
+      
+      etag = input.etags[0];
```

Comment:
Shouldn't we test this against the `etag` returns from the S3 HEAD request? and report InputError if the two don't match?

Reply:
That assumes eventual consistency is working as expected, and that no one has done an intermediate GET/HEAD before our HEAD which could cause a badly-cached resource to be stored in the S3 system.

Reply:
> That assumes eventual consistency is working as expected

No eventual consistency involved here, we rely on read-after-write consistency for first PUT.

> that no one has done an intermediate GET/HEAD before our HEAD which could cause a badly-cached resource to be stored in the S3 system.

Elaborate? I don't see how anything like that can change the etag.

Plausibly, if you had two PUT requests where the last one changes the etag.
Afaik, etags are content based and can't be changed by uploading the same thing again, am I wrong here?
Also we sign everything in the headers for the requests so two different resources cannot be uploaded.

Please outline the cases in which the etag can be made inconsistent by worker...

Reply:
@jhford ^